py/dml/types_test.py

```diff
@@ -0,0 +1,282 @@
+# © 2024 Intel Corporation
+# SPDX-License-Identifier: MPL-2.0
+
+from dml.types import *
+import dml.globals
+from dml import ctree
+from dml import expr
+from dml import traits
+from dml import logging
+
+import unittest
+
+class TestClone(unittest.TestCase):
+    def test(self):
+        # types which support clone
+        typ0 = TVoid()
+        for typ in (TVoid(),
+                    TNamed("int"),
+                    TBool(),
+                    TInt(8, False, {}),
+                    TEndianInt(8, False, 'big-endian', {}),
+                    TFloat("a"),
+                    TArray(typ0, ctree.mkIntegerLiteral(0, 2)),
+                    TPtr(typ0),
+                    TVector(typ0),
+                    TTrait(object()),
+                    TStruct({"name": TInt(32, False)}),
+                    TLayout("big-endian", {}),
+                    TDevice("a")):
+            typ_clone = typ.clone()
+            self.assertTrue(
+                realtype(typ_clone).eq(realtype(typ)))
+            self.assertTrue(
+                realtype(typ).eq(realtype(typ_clone)))
+            typ_clone.const = True
+            self.assertFalse(typ.const)
+            typ = typ_clone.clone()
+            self.assertTrue(typ.const)
```

Comment:
Note this addition to the clone test. This catches the previously bad `TVoid` and `TBool` implementations.